kissim/comparison/tree.py

```diff
@@ -0,0 +1,309 @@
+        Distance matrix on which clustering is based.
+    outputfile : str or pathlib.Path
+        Path to kinase tree file (TREE file) in Newick format.
+    """
+
+    outputfile = Path(outputfile)
+
+    logger.info(f"Writing resulting tree to {outputfile}")
+    newick = ""
+    newick = _get_newick(tree, newick, tree.dist, list(distance_matrix), node_means)
+    with open(outputfile, "w") as f:
+        f.write(newick)
+
+
+def _to_kinase_annotation(distance_matrix, outputfile):
+    """
+    Save kinase annotations to file used for FigTree.
+
+    Parameters
+    ----------
+    distance_matrix : pandas.DataFrame
+        Distance matrix on which clustering is based.
+    outputfile : str or pathlib.Path
+        Path to kinase annotation file (CSV file) in FigTree format.
+    """
+
+    outputfile = Path(outputfile)
+
+    logger.info(f"Writing resulting kinase annotation to {outputfile}")
+
+    # Get kinase names from matrix
+    kinase_names = distance_matrix.columns.to_list()
+
+    # Query KLIFS for kinase details
+    klifs_session = klifs.setup_remote()
+    kinases = klifs_session.kinases.by_kinase_name(kinase_names)
+    kinases = kinases[kinases["species.klifs"] == "Human"]
+    kinases = kinases[["kinase.klifs_name", "kinase.family", "kinase.group"]]
+
+    # Save to file
+    kinases.to_csv(outputfile, sep="\t", index=False)
+
+
+def _get_node_means(tree, distance_matrix, node_means, similarity=True):
+    """
+    Calculating and assign the mean similarity (default) or mean distance for tree branches.
+
+    Parameters
+    ----------
+    tree : scipy.cluster.hierarchy.ClusterNode
+        Clustering tree.
+    distance_matrix : pandas.DataFrame
+        Distance matrix on which clustering is based.
+    node_means : dict of int: float
+        Mean distance (value) for each node index (key).
+    similarity : bool
+        If `True`, convert distance matrix into similarity matrix using
+        `1 - distance_matrix / distance_matrix.max().max()`, else use distance matrix directly.
+
+    Returns
+    -------
+    node_means : dict of int: float
+        Mean similarities (value) for each node index (key).
+        Note: The return value is the populated input `node_means` object.
+    """
+
+    # Calculate mean similarities or mean distances?
+    if similarity:
+        matrix = (1 - distance_matrix / distance_matrix.max().max()).copy()
+        leaf_mean = 1
+    else:
+        matrix = distance_matrix.copy()
+        leaf_mean = 0
+
+    # Set mean for root node
+    if tree.id not in node_means:
+        # Get full matrix (root includes full dataset)
+        matrix_all_clusters = matrix.values
+        # Get matrix values from lower triangle without diagonal
+        values_all_clusters = matrix_all_clusters[
+            np.tril_indices(matrix_all_clusters.shape[0], -1)
+        ]
+        # Calculate mean of full dataset
+        node_means[tree.id] = np.average(values_all_clusters)
```

Comment:
I also added here the root node mean similarity, this was a fixed value before (1). 

If this does not make any sense, I am happy to delete this again (for the newick format it does not make a change, since we do not write the root mean to file).